File: src/types/node.ts
// src/types/node.ts

/**
 * Unique identifier for a node within a workflow.
 * Format: nanoid or uuid, opaque to business logic.
 */
export type NodeId = string & { readonly __brand: "NodeId" };

export type NodeType =
  | "agent" // Executes an AI agent
  | "tool" // Executes a deterministic tool/function
  | "input" // Workflow entry point (user-provided data)
  | "output"; // Workflow exit point (final result)

/**
 * Visual position on the canvas. Not relevant to execution.
 */
export interface NodePosition {
  readonly x: number;
  readonly y: number;
}

/**
 * Node-specific configuration. Varies by node type.
 */
export interface NodeConfig {
  readonly agentId?: string;
  readonly toolId?: string;
  readonly parameters?: Record<string, unknown>;
}

/**
 * A single node in the workflow DAG.
 */
export interface Node {
  readonly id: NodeId;
  readonly type: NodeType;
  readonly label: string;
  readonly position: NodePosition;
  readonly config: NodeConfig;
}

File: apps/api/src/agentforge_api/models/node.py
# apps/api/src/agentforge_api/models/node.py

"""Node domain models."""

from enum import StrEnum
from typing import Annotated, Any, NewType

from pydantic import BaseModel, Field

# Branded type equivalents using NewType
NodeId = NewType("NodeId", str)


class NodeType(StrEnum):
    """The category of node, determining its execution behavior."""

    AGENT = "agent"  # Executes an AI agent
    TOOL = "tool"  # Executes a deterministic tool/function
    INPUT = "input"  # Workflow entry point (user-provided data)
    OUTPUT = "output"  # Workflow exit point (final result)


class NodePosition(BaseModel, frozen=True):
    """Visual position on the canvas. Not relevant to execution."""

    x: float
    y: float


class NodeConfig(BaseModel, frozen=True):
    """
    Node-specific configuration. Varies by node type.
    Will be refined with discriminated unions later.
    """

    agent_id: str | None = None
    tool_id: str | None = None
    parameters: dict[str, Any] = Field(default_factory=dict)


class Node(BaseModel, frozen=True):
    """A single node in the workflow DAG."""

    id: Annotated[str, Field(description="Unique node identifier")]
    type: NodeType
    label: str
    position: NodePosition
    config: NodeConfig = Field(default_factory=NodeConfig)

    @property
    def node_id(self) -> NodeId:
        """Return typed NodeId."""
        return NodeId(self.id)
